server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java

```diff
@@ -0,0 +1,286 @@
+        MailQueue.MailQueueItem mailQueueItem2 = getMailQueue().deQueue();
+        mailQueueItem1.done(true);
+        mailQueueItem2.done(true);
+        assertThat(mailQueueItem1.getMail().getName()).isEqualTo("name1");
+        assertThat(mailQueueItem2.getMail().getName()).isEqualTo("name2");
+    }
+
+
+    @Test
+    default void dequeueCouldBeInterleavingWithOutOfOrderAck() throws Exception {
+        getMailQueue().enQueue(defaultMail()
+            .name("name1")
+            .build());
+        getMailQueue().enQueue(defaultMail()
+            .name("name2")
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem1 = getMailQueue().deQueue();
+        MailQueue.MailQueueItem mailQueueItem2 = getMailQueue().deQueue();
+        mailQueueItem2.done(true);
+        mailQueueItem1.done(true);
+        assertThat(mailQueueItem1.getMail().getName()).isEqualTo("name1");
+        assertThat(mailQueueItem2.getMail().getName()).isEqualTo("name2");
+    }
+
+    @Test
+    default void dequeueShouldAllowRetrieveFailItems() throws Exception {
+        getMailQueue().enQueue(defaultMail()
+            .name("name1")
+            .build());
+        getMailQueue().enQueue(defaultMail()
+            .name("name1")
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem1 = getMailQueue().deQueue();
+        mailQueueItem1.done(false);
+        MailQueue.MailQueueItem mailQueueItem2 = getMailQueue().deQueue();
+        mailQueueItem2.done(true);
+        assertThat(mailQueueItem1.getMail().getName()).isEqualTo("name1");
+        assertThat(mailQueueItem2.getMail().getName()).isEqualTo("name1");
+    }
+
+    @Test
+    default void dequeueShouldNotReturnInProcessingEmails(ExecutorService executorService) throws Exception {
+        getMailQueue().enQueue(defaultMail()
+            .name("name")
+            .build());
+
+        getMailQueue().deQueue();
+
+        Future<?> future = executorService.submit(Throwing.runnable(() -> getMailQueue().deQueue()));
+        assertThatThrownBy(() -> future.get(2, TimeUnit.SECONDS))
+            .isInstanceOf(TimeoutException.class);
+    }
+
+    @Test
+    default void deQueueShouldFreezeWhenNoMail(ExecutorService executorService) throws Exception {
+        Future<?> future = executorService.submit(Throwing.runnable(() -> getMailQueue().deQueue()));
+
+        assertThatThrownBy(() -> future.get(2, TimeUnit.SECONDS))
+            .isInstanceOf(TimeoutException.class);
+    }
+
+    @Test
+    default void deQueueShouldWaitForAMailToBeEnqueued(ExecutorService executorService) throws Exception {
+        Mail mail = defaultMail()
+            .name("name")
+            .build();
+        Future<MailQueue.MailQueueItem> tryDequeue = executorService.submit(() -> getMailQueue().deQueue());
+        executorService.submit(Throwing.runnable(() -> getMailQueue().enQueue(mail)));
+
+        assertThat(tryDequeue.get().getMail().getName()).isEqualTo("name");
+    }
+
```

Comment:
Other tests cases ideas :
are failed transactions pushed back as next element to be dequeued or elsewhere ?
what happens when you call `done` in various broken order with various values ?
(not sure it doesn't exist yet) do you need to ack pending elements to consume new ones ?

Reply:
```
what happens when you call done in various broken order with various values ?
```

Sorry, I don't get this one?

Reply:
```
do you need to ack pending elements to consume new ones ?
```

That's what `dequeueCouldBeInterleaving` demonstrate.

Reply:
> what happens when you call done in various broken order with various values ?

* queue 3 messages, dequeue the 3, call done(false) on 3, then 2, then 1. Check message order in the queue.

* queue 3 messages, dequeue the 3, call done(true) on 3, then done(false) on 2 and 1. Check message order in the queue.

etc.